run/events_pubsub/main_test.go

```diff
@@ -15,43 +15,85 @@
 package main
 
 import (
-	"context"
-	"encoding/json"
+	"encoding/base64"
+	"fmt"
+	"io/ioutil"
 	"log"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"strings"
 	"testing"
-	"time"
-
-	cloudevents "github.com/cloudevents/sdk-go/v2"
 )
 
-func TestHelloPubSubCloudEvent(t *testing.T) {
-	pubsubEvent := &PubSub{
-		Message: PubSubMessage{
-			Data: []byte("foo"),
-			ID:   "id",
+func TestHelloEventsPubSubErrors(t *testing.T) {
+	tests := []struct {
+		name    string
+		message string
+	}{
+		{
+			name:    "no_payload",
+			message: "",
+		},
+		{
+			name:    "not_base64",
+			message: `{"message":{"data":"Gopher","id":"test-123"}}`,
 		},
-		Subscription: "sub",
-	}
-	data, err := json.Marshal(pubsubEvent)
-	if err != nil {
-		log.Printf("json.Marshal: %v", err)
 	}
-	ce := &cloudevents.Event{
-		Context: cloudevents.EventContextV1{
-			ID:              "321-CBA",
-			Type:            "unit.test.client.response",
-			Time:            &cloudevents.Timestamp{Time: time.Now()},
-			Source:          *cloudevents.ParseURIRef("/unit/test/client"),
-			DataContentType: cloudevents.StringOfApplicationJSON(),
-		}.AsV1(),
-		DataEncoded: data,
+	for _, test := range tests {
+		payload := strings.NewReader(test.message)
+		req := httptest.NewRequest("POST", "/", payload)
+		rr := httptest.NewRecorder()
+
+		HelloEventsPubSub(rr, req)
+
+		if code := rr.Result().StatusCode; code != http.StatusBadRequest {
+			t.Errorf("HelloEventsPubSub(%q): got (%q), want (%q)", test.name, code, http.StatusBadRequest)
+		}
 	}
+}
 
-	got, e := HelloPubSub(context.Background(), *ce)
-	if e != nil {
-		t.Errorf("HelloPubSub: %q", e)
+func TestHelloEventsPubSub(t *testing.T) {
+	tests := []struct {
+		data string
+		want string
+		ID   string
+	}{
+		{want: "Hello, World! ID: \n"},
+		{data: "Go", want: "Hello, Go! ID: 1234\n", ID: "1234"},
 	}
-	if want := "Hello, foo! ID: 321-CBA"; got != want {
-		t.Errorf("HelloPubSub: got %q, want %q", got, want)
+	for _, test := range tests {
+		r, w, _ := os.Pipe()
+		log.SetOutput(w)
+		originalFlags := log.Flags()
+		log.SetFlags(log.Flags() &^ (log.Ldate | log.Ltime))
+
+		payload := strings.NewReader("{}")
+		if test.data != "" {
+			encoded := base64.StdEncoding.EncodeToString([]byte(test.data))
+			jsonStr := fmt.Sprintf(`{"message":{"data":"%s"}}`, encoded)
+			payload = strings.NewReader(jsonStr)
+		}
+		req := httptest.NewRequest("POST", "/", payload)
+		req.Header.Set("ce-id", test.ID)
+
+		rr := httptest.NewRecorder()
+		HelloEventsPubSub(rr, req)
+
+		w.Close()
+		log.SetOutput(os.Stderr)
+		log.SetFlags(originalFlags)
```

Comment:
Can update these outside the loop, then `defer` these calls so they're right next to the original modifications.

Reply:
I was able to set the flags outside of the loop.

I was not able to successfully update the output outside of the loop. I believe some of this logic is common with other files. The argument for setting this inside or outside of the loop could be argued both ways.

Reply:
Sorry, I'll need more guidance if we want to go this route. I've tried a few things.

This came from another test originally:
https://github.com/GoogleCloudPlatform/golang-samples/blob/a699cc86997175aa096113486151599df8948a10/functions/helloworld/hello_pubsub_test.go

Reply:
Figured it out.